pkg/cortex/modules.go

```diff
@@ -293,11 +293,15 @@ func (t *Cortex) initIngester(cfg *Config) (err error) {
 	cfg.Ingester.TSDBConfig = cfg.TSDB
 	cfg.Ingester.ShardByAllLabels = cfg.Distributor.ShardByAllLabels
 
-	t.ingester, err = ingester.New(cfg.Ingester, cfg.IngesterClient, t.overrides, t.store, prometheus.DefaultRegisterer)
-	if err != nil {
-		return
+	var errfut *ingester.ErrorFuture
+	t.ingester, errfut = ingester.New(cfg.Ingester, cfg.IngesterClient, t.overrides, t.store, prometheus.DefaultRegisterer)
+	if done, err := errfut.Get(); done && err != nil {
```

Comment:
Futures feel needlessly complicated here, and an anti-pattern with Go.  Concurrency should be left to the caller, so I believe the module init function should call `ingester.New` in a goroutine that waits for an error response.

Reply:
Hmm, that is a valid point. Back to drawing board.

Reply:
However, by contract the ingester should be populated in the `Cortex` struct once the `initIngester()` terminates.